src/os/bluestore/BlueStore.cc

```diff
@@ -698,48 +702,103 @@ void BlueStore::TwoQCache::trim(uint64_t onode_max, uint64_t buffer_max)
   _audit("trim start");
 
   // buffers
-  uint64_t num_buffers = buffer_hot.size() + buffer_warm_in.size();
-  if (num_buffers) {
-    // for simplicity, we assume buffer sizes are uniform across hot
-    // and warm_in.  FIXME?
-    uint64_t avg_buffer_size = buffer_bytes / num_buffers;
-    uint64_t kin = buffer_max / avg_buffer_size / 2;
-    uint64_t khot = kin;
-    uint64_t kout = buffer_max / avg_buffer_size;
-    // if hot is small, give back to warm_in.
-    if (buffer_hot.size() < kin) {
-      kin += kin - buffer_hot.size();
-    }
-    dout(20) << __func__ << " num_buffers " << num_buffers
-	     << " (hot " << buffer_hot.size()
-	     << " + warm_in " << buffer_warm_in.size() << ")"
-	     << " avg size " << avg_buffer_size
-	     << " warm_out " << buffer_warm_out.size()
-	     << " khot " << khot << " kin " << kin << " kout " << kout
-	     << dendl;
-    while (buffer_hot.size() > khot) {
-      Buffer *b = &*buffer_hot.rbegin();
-      assert(b->is_clean());
-      dout(20) << __func__ << " buffer_hot rm " << *b << dendl;
-      b->space->_rm_buffer(b);
-    }
-    while (buffer_warm_in.size() > kin) {
-      Buffer *b = &*buffer_warm_in.rbegin();
+  // first adjust clean(warm_in and hot) buffer lists
+  if (buffer_bytes > buffer_max) {
+
+    int64_t to_evict_bytes = buffer_bytes - buffer_max;
```

Comment:
It looks like this misses the key part of 2Q that sizes the warm_in vs hot lists.  It looks like this will trim everything from warm_in until it is empty before even touching hot items?